src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -324,6 +324,15 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
                         : this.ProjectState;
 
                     compilationWithoutGeneratedDocuments = inProgressState.CompilationWithoutGeneratedDocuments;
+
+                    // Parse the generated documents in parallel.
+                    Parallel.ForEach(
+                        generatorInfo.Documents.States.Values,
+                        new ParallelOptions { CancellationToken = cancellationToken },
+                        state => state.GetSyntaxTree(cancellationToken));
```

Comment:
This worries me if potentially Parallel.ForEach() were to try to start running stuff on other threads and end up getting blocked there. But if the framework is smart enough to keep using our own thread which potentially might do all the work then this isn't any worse than it is now.

You hit a more general point that it seems we're actually parsing files here, so I wonder if this another place we can defer this work until later like we did for the regular document. But you've just highlighted a problem I didn't appreciate until now.

Reply:
> FreezePartialState

yup.  that's one of the things i was exploring.